build_tools/github_actions/promote_wheels_based_on_policy.py

```diff
@@ -25,8 +25,8 @@ def determine_upload_flag(
     if build_result != "success":
         print("::warning::Build failed. Skipping upload.")
 
-    # 2) Else if there was a test runner AND tests failed or were skipped → upload=false
-    elif test_runs_on and (test_result in ["failure", "skipped"]):
+    # 2) Else if there was a test runner AND tests failed, skipped or were cancelled → upload=false
+    elif test_runs_on and (test_result in ["failure", "skipped", "cancelled"]):
         print(
             "::warning::Tests failed or were skipped (runner present). Skipping upload."
         )
```

Comment:
How about checking `test_result != "success"`? You could instead add a new `elif` for cancelled. I mainly want the condition to be easy to read.

```suggestion
    # 2) Else if there was a test runner AND tests did not succeed → upload=false
    elif test_runs_on and (test_result != "success"):
        print(
            "::warning::Tests were not successful (runner present). Skipping upload."
        )
```

Reply:
Oh sorry, clarifying: I'd _either_ do this:

```python
    # 2) Else if there was a test runner AND tests did not succeed → upload=false
    elif test_runs_on and (test_result != "success"):
        print(
            "::warning::Tests were not successful (runner present). Skipping upload."
        )
```

OR this:

```python
    # 2) Else if there was a test runner AND tests failed or were skipped → upload=false
    elif test_runs_on and (test_result in ["failure", "skipped"]):
        print(
            "::warning::Tests failed or were skipped (runner present). Skipping upload."
        )

    # 3) Else if there was a test runner AND tests were cancelled → upload=false
    elif test_runs_on and (test_result == "cancelled"):
        print(
            "::warning::Tests were cancelled (runner present). Skipping upload."
        )
```

I prefer just the first one, as long as it catches all cases we care about.

In fact, you could also rework the logging like so:

```python
    # 2) Else if there was a test runner AND tests did not succeed → upload=false
    elif test_runs_on and (test_result != "success"):
        print(
            f"::warning::Runner present and tests were not successful (test_result: {test_result}). Skipping upload."
        )
```

Reply:
`    # 2) Else if there was a test runner AND tests did not succeed → upload=false
    elif test_runs_on and (test_result != "success"):
        print(
            f"::warning::Runner present and tests were not successful (test_result: {test_result}). Skipping upload."
        )`
        
 This condition covers all cases we are covering above. Will adopt this. Thanks